nuxeo-services/login/nuxeo-platform-login-jwt/src/main/java/org/nuxeo/ecm/jwt/JWTServiceImpl.java

```diff
@@ -0,0 +1,323 @@
+            } else if (value instanceof String[]) {
+                builder.withArrayClaim(name, (String[]) value);
+            } else {
+                throw new NuxeoException("Unknown claim type: " + value);
+            }
+            return this;
+        }
+
+        @Override
+        public String build() {
+            try {
+                Algorithm algorithm = getAlgorithm();
+                if (algorithm == null) {
+                    throw new NuxeoException("JWTService secret not configured");
+                }
+                return builder.sign(algorithm);
+            } catch (JWTCreationException e) {
+                throw new NuxeoException(e);
+            }
+        }
+    }
+
+    protected void builderWithClaim(Builder builder, String name, Object value) {
+        if (value instanceof Boolean) {
+            builder.withClaim(name, (Boolean) value);
+        } else if (value instanceof Date) {
+            builder.withClaim(name, (Date) value);
+        } else if (value instanceof Double) {
+            builder.withClaim(name, (Double) value);
+        } else if (value instanceof Integer) {
+            builder.withClaim(name, (Integer) value);
+        } else if (value instanceof Long) {
+            builder.withClaim(name, (Long) value);
+        } else if (value instanceof String) {
+            builder.withClaim(name, (String) value);
+        } else if (value instanceof Integer[]) {
+            builder.withArrayClaim(name, (Integer[]) value);
+        } else if (value instanceof Long[]) {
+            builder.withArrayClaim(name, (Long[]) value);
+        } else if (value instanceof String[]) {
+            builder.withArrayClaim(name, (String[]) value);
+        } else {
+            throw new NuxeoException("Unknown claim type: " + value);
+        }
+    }
+
+    @Override
+    public Map<String, Object> verifyToken(String token) {
+        Objects.requireNonNull(token);
+        Algorithm algorithm = getAlgorithm();
+        if (algorithm == null) {
+            log.debug("secret not configured, cannot verify token");
+            return null; // no secret
+        }
+        JWTVerifier verifier = JWT.require(algorithm) //
+                                  .withIssuer(NUXEO_ISSUER)
+                                  .build();
+        DecodedJWT jwt;
+        try {
+            jwt = verifier.verify(token);
+        } catch (JWTVerificationException e) {
+            if (log.isTraceEnabled()) {
+                log.trace("token verification failed: " + e.toString());
+            }
+            return null; // invalid
+        }
+        Object payload = getFieldValue(jwt, "payload"); // com.auth0.jwt.impl.PayloadImpl
+        Map<String, JsonNode> tree = getFieldValue(payload, "tree");
+        Map<String, Object> claims = new HashMap<>();
+        for (Entry<String, JsonNode> es : tree.entrySet()) {
+            claims.put(es.getKey(), nodeToValue(es.getValue()));
+        }
```

Comment:
Could be replaced with `Map<String, Object> claims = tree.entrySet().stream().collect(Collectors.toMap(Entry::getKey, e -> nodeToValue(e.getValue())))`

Reply:
👍